ethcore/src/blockchain/blockchain.rs

```diff
@@ -612,14 +611,22 @@ impl BlockChain {
 					.add_bloom(&header.log_bloom(), header.number() as usize)
 			},
 			BlockLocation::BranchBecomingCanonChain { ref ancestor, ref route } => {
-				let ancestor_number = self.block_number(ancestor).unwrap();
+				let ancestor_number = try!(self.block_number(ancestor).ok_or(FatalError::MissingBlockDetails { hash: ancestor.clone() }));
 				let start_number = ancestor_number + 1;
 
-				let mut blooms: Vec<H2048> = route.iter()
-					.map(|hash| self.block(hash).unwrap())
-					.map(|bytes| BlockView::new(&bytes).header_view().log_bloom())
+				let blocks: Vec<(H256, Option<Bytes>)> = route.iter()
+					.map(|hash| (hash.clone(), self.block(hash)))
 					.collect();
 
+				let mut blooms: Vec<H2048> = Vec::new();
+
+				for block in blocks.into_iter() {
```

Comment:
@gavofyork Problem is with this line `.map(|hash| self.block(hash).unwrap())` we would like to somehow propagate error information (including `hash` as part of error message). I spent some time with @debris looking at this and it was one reasonable solution. I don't know enough rust to do this in more functional way, but maybe you could suggest something?